src/util.cc

```diff
@@ -2428,38 +2462,56 @@ string json_escape_utf8(const string& val)
 		if ( ( ch >= 32 && ch <= 127 ) ||
 		       ch == '\b' || ch == '\f' || ch == '\n' || ch == '\r' || ch == '\t' )
 			{
-			result.push_back(ch);
+			if ( ! found_bad )
+				utf_result.push_back(ch);
+
+			escaped_result.push_back(ch);
 			++idx;
 			continue;
 			}
-		else if ( ch >= 0 && ch < 32 )
+		else if ( found_bad )
 			{
-			result.append(json_escape_byte(ch));
+			// If we already found a bad UTF8 character (see check_ok_utf8) just insert the bytes
+			// as escaped characters into the escaped result and move on.
+			escaped_result.append(json_escape_byte(ch));
 			++idx;
 			continue;
 			}
 
-		// Find out how long the next character should be.
-		unsigned int char_size = getNumBytesForUTF8(ch);
-
-		// If it says that it's a single character or it's not an valid string UTF8 sequence, insert
-		// the one escaped byte into the string, step forward one, and go to the next character.
-		if ( char_size == 0 || idx+char_size > val_size || isLegalUTF8Sequence(val_data+idx, val_data+idx+char_size) == 0 )
+		// If we haven't found a bad UTF-8 character yet, check to see if the next one starts a
+		// UTF-8 character. If not, we'll mark that we're on a bad result. Otherwise we'll go
+		// ahead and insert this character and continue.
+		if ( ! found_bad )
 			{
-			result.append(json_escape_byte(ch));
-			++idx;
-			continue;
-			}
+			// Find out how long the next character should be.
+			unsigned int char_size = getNumBytesForUTF8(ch);
 
-		result.append(val, idx, char_size);
-		idx += char_size;
+			// If it says that it's a single character or it's not an valid string UTF8 sequence, insert
+			// the one escaped byte into the string, step forward one, and go to the next character.
+			if ( char_size == 0 || idx+char_size > val_size ||
```

Comment:
Is `getNumBytesForUTF8(ch)` returning zero if it's a single character?

Reply:
Yes. It's possible for it to return 0. See https://github.com/zeek/zeek/blob/master/src/ConvertUTF.c#L117-L133.

Reply:
Oh, but look at this:

 ```
unsigned getNumBytesForUTF8(UTF8 first) {
  return trailingBytesForUTF8[first] + 1;
}
```

 That will always return >= 1.

Reply:
Ah, you are correct. Now I'm confused though, because according to that table I linked, characters 128-192 are considered a single byte. That means we wouldn't ever insert those characters (things like the copyright symbol, etc). It's probably ok to just take that check out entirely, since it hasn't been doing anything anyways.

That said, there's a bunch of non-printable characters in that second block of 127 characters that we should skip (see https://codepoints.net/latin-1_supplement).

Reply:
Ah, https://codepoints.net/U+00C7 shows that the UTF-8 version of at least one of those characters is actually a two-byte sequence (and that's true for all of them). I wonder why that function returns `1` for a bunch of them.